### DISCUSSION TITLE
Logging to multiple syslog facility

### DISCUSSION THREAD

**Discussion by jacoson** [gabime/spdlog#2174] [Q&A]
Hello everyone,

Can you please help me about how I can log to multiple syslog facility?

I've created 2 seperate loggers with same sink (and with different syslog facility settings)

(I am planning to use multiple sinks this is why the structure below has sinks vector)

```
// Logger-1 
{
    std::string logger1 = "logger1";
    std::vector<spdlog::sink_ptr> sinks;
    std::string ident = "log-to-var-log-messages";
    auto syslog_sink = std::make_shared<spdlog::sinks::syslog_sink_mt>(ident, LOG_CONS | LOG_NDELAY, LOG_USER, false);
    sinks.push_back(syslog_sink);
    auto logger = std::make_shared<spdlog::async_logger>(logger1, sinks.begin(), sinks.end(), 
                                                             spdlog::thread_pool(), spdlog::async_overflow_policy::block);
    logger->set_level(spdlog::level::trace);
    logger->flush_on(spdlog::level::trace);
    spdlog::register_logger(logger);
}
```

```
// Logger-2
{
    std::string logger2 = "logger2";
    std::vector<spdlog::sink_ptr> sinks;
    std::string ident = "log-to-var-log-secure";
    auto syslog_sink = std::make_shared<spdlog::sinks::syslog_sink_mt>(ident, LOG_CONS | LOG_NDELAY, LOG_AUTHPRIV, false);
    sinks.push_back(syslog_sink);
    auto logger = std::make_shared<spdlog::async_logger>(logger2, sinks.begin(), sinks.end(), 
                                                             spdlog::thread_pool(), spdlog::async_overflow_policy::block);
    logger->set_level(spdlog::level::trace);
    logger->flush_on(spdlog::level::trace);
    spdlog::register_logger(logger);
}
```

This is my macro

`#define LOG(loggerId, LEVEL, ...) spdlog::get(loggerId)->log(spdlog::source_loc{__FILE__, __LINE__, SPDLOG_FUNCTION}, LEVEL, "{}", fmt::sprintf(fmt, args...)`

when I call these lines, it logs these 2 lines into /var/log/secure only ??? 

```
LOG("logger1", INFO, "this should be displayed in /var/log/messages");
LOG("logger2", INFO, "this should be displayed in /var/log/secure");

```

**Comment by tt4g** [comment]
Is syslog (rsyslog) configured to output AUTHPRIV to `/var/log/secure`?

**Comment by jacoson** [comment]
Thanks for your reply tt4g, 
Yes, actually no problem with AUTHPRIV, I can see the logs in /var/log/secure but I think (as far as I understood), desired output should be 1 log into the /var/log/messages and 1 log into the /var/log/secure.
But I can not see any log in /var/log/messages but 2 logs in /var/log/secure

**Reply by tt4g** [reply]
Sorry, I misunderstood that all the logs are output to `/var/log/messages`.
Does using the `logger` command output to `/var/log/secure`?

Example:

```bash
$ logger -ip user.notice "this should be displayed in /var/log/messages"
```

**Reply by jacoson** [reply]
No, this command writes into /var/log/messages as expected I guess. 
It looks like I can use only 1 syslog sink and if I define a logger with LOG_AUTHPRIV, it disables the syslog logger with LOG_USER facility and writes every log into /var/log/secure. 
On the other hand, if I initialize only 1 logger with syslog sink (LOG_USER), I can write every log into /var/log/messages. 
But I can't use both syslog loggers with LOG_USER and LOG_AUTHPRIV.

**Comment by tt4g** [comment]
I found calling `openlog()` in the `spdlog::sinks::syslog_sink` constructor (https://github.com/gabime/spdlog/blob/v1.9.2/include/spdlog/sinks/syslog_sink.h#L36).

Since `openlog()` sets the facility when the process calls `syslog()`, it is not possible to specify a different facility for each `spdlog::sinks::syslog_sink`.

Apparently, facility can be specified with the first argument level of `syslog()` (https://stackoverflow.com/a/24008360), but other `openlog()` options are not supported.

Unfortunately, spdlog v1.9.2 does not support using multiple `spdlog::sinks::syslog_sink`.
